### PR TITLE
Nonlinear regression simulations for existing split criteria

### PR DESCRIPTION
**Pull request by morgsmss7**
#### Reference Issues/PRs

Fixes Issue [16370](https://github.com/scikit-learn/scikit-learn/issues/16370) in scikit-learn. Also see Issue [2](https://github.com/NeuroDataDesign/tealeaf/issues/2) in tealeaf. 

#### What does this implement/fix? Explain your changes.
This PR adds simulations and plots that show how split criteria compare on several nonlinear regression simulations including sinusoidal, logarithmic, multiplicative, and independence. There is not much information on scikit-learn's documentation about how to go about choosing which to use (mse, mae, or friedman mse) for the `criterion` parameter. This example demonstrates how to go about finding differences and shows that it may not always matter which criterion is chosen. 

#### Any other comments?
This PR in `sklearn` will include these files:
```
sklearn/datasets/tests/test_samples_generator.py 
sklearn/datasets/samples_generator.py
sklearn/datasets/__init__.py
examples/ensemble/plot_random_forest_regression_criteria_comparison.py
examples/datasets/plot_nonlinear_regression_datasets.py 
doc/modules/classes.rst 
doc/datasets/index.rst
```

The other files that were changed for Vivek's PR will not be changed in sklearn.

### CODE REVIEW

**Comment by morgsmss7** [comment]
# Split Criteria Comparison Experiment and Results
For each simulation type (Logarithmic, Sine, Square, Multiplicative, Independence)
And For each split criterion (mse, mae, friedman mse)
1. Generate 30 noisy training sets (10 dimensions, 50 samples) 
2. Generate 1 noisy test set (10 dimensions, 1000 samples)
3. Train and evaluate using mse for all 30 training sets with random forests (500 trees) varying number of samples ( `np.arange(5, 51, 3)` )

nonlinearSimPlots
splitter_comparison_02_17

File: examples/ensemble/plot_random_forest_regression_criteria_comparison.py [hunk 1]
```diff
@@ -62,24 +63,65 @@ def _test_forest(X, y, regr):
...
 ###############################################################################
-def main(simulation_name, n_samples, criterion, n_dimensions, n_iter):
+def main(simulation_name, sim_data, n_samples, criterion, n_dimensions, n_iter):
     """Measure the performance of RandomForest under simulation conditions.
-
     Parameters
     ----------
     simulation_name : str
         Key from `simulations` dictionary.
+    sim_data: tuple
```
Comment (v715) [inline examples/ensemble/plot_random_forest_regression_criteria_comparison.py]: Make this a single line
`sim_data: tuple (X_train, y_train, X_test, and y_test`

File: examples/ensemble/plot_random_forest_regression_criteria_comparison.py [hunk 2]
```diff
@@ -62,24 +63,65 @@ def _test_forest(X, y, regr):
...
+        Contains X_train, y_train, X_test, and y_tests
+            X_train : np.array #TODO check this
+                All X training data for given simulation
+            y_train : np.array # TODO
+                All y training data for given simulation
+            X_test : np.array #TODO check this
+                All X testing data for given simulation
+            y_test : np.array # TODO
+                All y testing data for given simulation
```
Comment (eigenvivek) [inline examples/ensemble/plot_random_forest_regression_criteria_comparison.py]: What's with the `# TODO`?

Also might be nice to add the shape of the arrays as is done in `graspy` ([example](https://github.com/neurodata/graspy/blob/287bb762382aa00ae7d3d0702fcf53b71b989cc7/graspy/embed/omni.py#L108-L109)).

File: examples/ensemble/plot_random_forest_regression_criteria_comparison.py [hunk 3]
```diff
@@ -62,24 +63,65 @@ def _test_forest(X, y, regr):
...
-def main(simulation_name, n_samples, criterion, n_dimensions, n_iter):
+def main(simulation_name, sim_data, n_samples, criterion, n_dimensions, n_iter):
     """Measure the performance of RandomForest under simulation conditions.
-
     Parameters
     ----------
     simulation_name : str
         Key from `simulations` dictionary.
+    sim_data: tuple
+        Contains X_train, y_train, X_test, and y_tests
```
Comment (eigenvivek) [inline examples/ensemble/plot_random_forest_regression_criteria_comparison.py]: Make this a single line like `sim_data: tuple (X_train, y_train, X_test, and y_test)`

File: examples/ensemble/plot_random_forest_regression_criteria_comparison.py [hunk 4]
```diff
@@ -62,24 +63,65 @@ def _test_forest(X, y, regr):
...
+            X_test : np.array #TODO check this
+                All X testing data for given simulation
+            y_test : np.array # TODO
+                All y testing data for given simulation
     n_samples : int
         Number of training samples.
     criterion : string
         Split criterion used to train forest. Choose from
-        ("mse", "mae", "friedman_mse", "axis", "oblique").
+        ("mse", "mae", "friedman_mse").
```
Comment (eigenvivek) [inline examples/ensemble/plot_random_forest_regression_criteria_comparison.py]: Again consider changing this to something more stylistic like is in `graspy` ([example](https://github.com/neurodata/graspy/blob/287bb762382aa00ae7d3d0702fcf53b71b989cc7/graspy/embed/omni.py#L79))

File: examples/ensemble/plot_random_forest_regression_criteria_comparison.py [hunk 5]
```diff
@@ -133,50 +166,60 @@ def main(simulation_name, n_samples, criterion, n_dimensions, n_iter):
...
+    x="n_samples",
+    y="mse",
+    hue="criterion",
+    col="simulation",
+    kind="line",
+    data=df,
+    facet_kws={"sharey": False, "sharex": True},
+)
+plt.tight_layout()
+plt.savefig("splitter_comparison_02_17.png")
```
Comment (eigenvivek) [inline examples/ensemble/plot_random_forest_regression_criteria_comparison.py]: No need to save the plot!

File: examples/ensemble/plot_random_forest_regression_criteria_comparison.py [hunk 6]
```diff
@@ -133,50 +166,60 @@ def main(simulation_name, n_samples, criterion, n_dimensions, n_iter):
...
+# Plot the results
+sns.relplot(
+    x="n_samples",
+    y="mse",
+    hue="criterion",
+    col="simulation",
+    kind="line",
+    data=df,
+    facet_kws={"sharey": False, "sharex": True},
+)
```
Comment (eigenvivek) [inline examples/ensemble/plot_random_forest_regression_criteria_comparison.py:226]: Not sure if `scikit-learn` will like this because I don't believe `pandas` or `seaborn` are part of the default installation... but let's keep it in since it makes plotting so much easier!

**Review by eigenvivek — DISMISSED** [review]
Overall, it looks great! I made some minor comments asking for stylistic changes to the comments.

**Comment by j1c** [comment]
This looks great. I think you wanted to make a PR to real sklearn right? My only concern is that the current NDD master has bunch of changes from other people that would be merged in as well. For this, I think the best course of action is to make a new branch, fetch the latest sklearn, add in these two examples along with the data generation code. Then make the PR from that branch.
